Add Wrangler support for R2 storage classes

ocsfrank opened the pull request:
## What this PR solves / how to test

Not exempt from LDW, but I would like to get this reviewed before next week.

Fixes R2-1963

## Author has addressed the following

- Tests
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Changeset ([Changeset guidelines](https://github.com/cloudflare/workers-sdk/blob/main/CONTRIBUTING.md#changesets))
  - [ ] TODO (before merge)
  - [x] Included
  - [ ] Not necessary because:
- Public documentation
  - [ ] TODO (before merge)
  - [ ] Cloudflare docs PR(s): <https://github.com/cloudflare/cloudflare-docs/pull/>...
  - [x] Not necessary because: pre-release feature. Docs are planned for next week and will be retro-actively linked here @jonesphillip

sdnts commented on packages/wrangler/src/__tests__/r2.test.ts:
```diff
@@ -42,11 +42,12 @@ describe("r2", () => {
-			  wrangler r2 bucket sippy          Manage Sippy incremental migration on an R2 bucket
-			  wrangler r2 bucket notification   Manage event notifications for an R2 bucket
+			  wrangler r2 bucket create <name>             Create a new R2 bucket
+			  wrangler r2 bucket put-storage-class <name>  Update the default storage class of an existing R2 bucket
```
Have you considered making a `put` subcommand so we can perhaps nest other bucket updates under it?
`wrangler r2 bucket put storage-class <name>`

ocsfrank replied:
Good idea -- we indeed should.

ocsfrank replied:
Added as a subcommand.

petebacondarwin commented on packages/wrangler/src/__tests__/r2.test.ts:
```diff
@@ -196,12 +199,123 @@ describe("r2", () => {
+						async (request, response, context) => {
+							const { accountId } = request.params;
+							expect(accountId).toEqual("some-account-id");
+							expect(request.headers.get("cf-r2-storage-class")).toEqual(
```
I know we do this all over the place, but really we should not be using `expect` checks inside the mock service worker handlers.
Instead, we should simulate what the backend would do, which I assume would be some kind of 4xx response in the case that the storage class is invalid?

Then in these tests we can simply check that Wrangler does not error itself, which proves that it must have passed the correct header.
Equally we can check that Wrangler handles the error response correctly if the header is invalid.

ocsfrank replied:
Good point, going to update the test.

petebacondarwin commented on .changeset/unlucky-laws-fetch.md:
````diff
@@ -0,0 +1,15 @@
+
+```bash
+wrangler r2 bucket create ia-bucket -s InfrequentAccess
+wrangler r2 bucket put-storage-class my-bucket -s InfrequentAccess
````
Oh from looking at the code it should be:

```suggestion
wrangler r2 bucket put storage-class my-bucket -s InfrequentAccess
```

The extra noun "put" seems a bit off here. The object of the command is the `bucket` still and the operation is `update-storage-class`, no?

petebacondarwin replied:
So I would imagine this command should be:

```suggestion
wrangler r2 bucket update-storage-class my-bucket -s InfrequentAccess
```

ocsfrank replied:
Ah, sorry -- that's my bad. I updated the command itself, but not the changeset file. The reason for using `put` even though it's a `PATCH` request is because the action itself is called `PutBucketStorageClass` in the API to be consistent with other operations that update a bucket settings (`PutBucketLifecycleConfiguration`, `PutBucketCors`, etc.). So, this is why there's a `put` subcommand as we'd also like to expand Wrangler to include the other bucket actions.

ocsfrank replied:
Although, looking at this again, I do agree that `update` makes it much more intuitive from a user perspective as the operations **is** a `PATCH` request. Updating this to replace the `put` subcommand with `update` subcommand.

petebacondarwin commented on packages/wrangler/src/r2/helpers.ts:
```diff
@@ -48,19 +48,47 @@ export async function listR2Buckets(
+ * Update the default storage class to `storageClass` of a bucket with the given `bucketName`
+ * within the account given by `accountId`.
+ */
+export async function putR2BucketStorageClass(
```
`put...` seems like the wrong name here?
This is updating the storage class using a `PATCH` verb not `PUT`.
It feels like "put" is being used because it was copied over from when we actually "PUT" to the REST API?

ocsfrank replied:
It's because of the naming convention in the R2 API itself (https://github.com/cloudflare/workers-sdk/pull/5429#discussion_r1546741564).

ocsfrank commented:
Updated to using `update storage-class` instead of `put storage-class` as it's more intuitive. Also updated the tests to mimic the response from R2.

ocsfrank commented on packages/wrangler/src/__tests__/r2.test.ts:
```diff
@@ -850,6 +975,18 @@ describe("r2", () => {
 		`);
 			});
 
+			it("should upload R2 object with storage class to bucket", async () => {
```
Does this test run against real R2? If that's the case, we'll have to remove it as it's currently private preview.

ocsfrank replied:
Ah, it doesn't look like it -- it's using mock api and token. I think we're good.
